File: http/src/main/Plume/Http/Routes/Server/ActionsRoute.kt
package Plume.Http.Routes.Server

import Plume.Configuration.ConfigManager
import Plume.Database.DatabaseManager
import Plume.Http.RouteModule
import Plume.Http.Routes.errorJson
import Plume.Http.Routes.successJson

import Plume.Server.Manager
import Plume.Server.Extending.install
import Plume.Server.Extending.kill
import Plume.Server.Extending.sendCommand
import Plume.Server.Extending.start
import Plume.Types.*
import io.ktor.http.HttpStatusCode
import io.ktor.server.request.receiveNullable
import io.ktor.server.response.respond
import io.ktor.server.routing.Route
import io.ktor.server.routing.post
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlin.time.Duration.Companion.milliseconds

object ActionsRoute : RouteModule {
    // Pré-alocação de erros comuns para economizar RAM e evitar Garbage Collection sob spam
    private val ERR_INVALID_JSON = errorJson("Invalid JSON")
    private val ERR_INVALID_SERVER = errorJson("Invalid serverId")
    private val ERR_INVALID_UUID = errorJson("Invalid userUuid")
    private val ERR_INVALID_TOKEN = errorJson("Invalid token")
    private val ERR_MISSING_SERVER = errorJson("serverId is required")
    private val ERR_NOT_FOUND = errorJson("Server not found")
    private val ERR_INSTALLING = errorJson("server is installing")
    private val ERR_DB_NOT_FOUND = errorJson("Server not found in database")
    private val ERR_DISK_LIMIT = errorJson("Disk usage exceeds the limit")
    private val ERR_MISSING_FIELDS = errorJson("Missing required fields")
    private val ERR_NOT_STOPPED = errorJson("server not stopped")
    private val ERR_INVALID_ACTION = errorJson("Invalid action")
    private val SUCCESS_RESPONSE = successJson()

    // --- DTOs PARA LER O JSON DO FRONTEND ---
    @Serializable
    data class ActionPayloadCore(
        val installScript: String = "",
        val installImage: String = "",
        val installEntrypoint: String = "",
        val startupCommand: String = "",
        val startupScript: String = "",
        val dockerEntrypoint: String = "",
        val stopCommand: String = "",
        val configSystem: JsonElement? = null,
        val startupParser: JsonElement? = null,
        val rootAcess: Long? = null,
        val maintainable: Long? = null
    )

    @Serializable
    data class ActionPayload(
        val token: String = "",
        val serverId: String = "",
        val userUuid: String = "",
        val action: String = "",
        val command: String = "",

        val memory: Int = 0,
        val cpu: Int = 0,
        val disk: Long = 0,
        val image: String = "",
        val environment: JsonElement? = null,
        val primaryAllocation: AllocationData? = null,
        val additionalAllocation: List<AllocationData> = emptyList(),

        val core: ActionPayloadCore = ActionPayloadCore()
    )

    private val serverActionScope = CoroutineScope(Dispatchers.IO)

    override fun register(route: Route) {
        route.post("/servers/action") {
            // Tenta parsear o body
            val body = try {
                call.receiveNullable<ActionPayload>()
            } catch (_: Exception) {
                null
            }

            if (body == null) {
                call.respond(HttpStatusCode.BadRequest, ERR_INVALID_JSON)
                return@post
            }

            // Validações de Entrada Rigorosas
            if (!ConfigManager.validateServerID(body.serverId)) {
                call.respond(HttpStatusCode.BadRequest, ERR_INVALID_SERVER)
                return@post
            }

            if (!ConfigManager.validateUUID(body.userUuid)) {
                call.respond(HttpStatusCode.BadRequest, ERR_INVALID_UUID)
                return@post
            }


            if (body.token != ConfigManager.globalConfig.remote.token) {
                call.respond(HttpStatusCode.Forbidden, ERR_INVALID_TOKEN)
                return@post
            }

            if (body.serverId.isEmpty()) {
                call.respond(HttpStatusCode.BadRequest, ERR_MISSING_SERVER)
                return@post
            }

            val srv = Manager.get(body.serverId)
            if (srv == null) {
                call.respond(HttpStatusCode.NotFound, ERR_NOT_FOUND)
                return@post
            }

            // Utiliza equals(ignoreCase) ao invés de alocar nova String com .lowercase()
            val currentStatus = srv.getStatus()
            if (currentStatus.equals("installing", ignoreCase = true)) {
                srv.emitLive("error", "O servidor está atualmente em instalação. Tente novamente mais tarde.")
                call.respond(HttpStatusCode.BadRequest, ERR_INSTALLING)
                return@post
            }

            val serverDb = DatabaseManager.db.getServer(body.serverId)
            if (serverDb == null) {
                call.respond(HttpStatusCode.NotFound, ERR_DB_NOT_FOUND)
                return@post
            }

            // Só executa o I/O pesado de gravar no banco se a configuração realmente mudou
            if (serverDb.disk != body.disk) {
                DatabaseManager.db.saveServer(serverDb.copy(disk = body.disk))
                srv.diskLimitMb = body.disk // Atualiza na memória
            }

            val action = body.action
            val isStopOrKill = action == "kill" || action == "stop"

            // Evita processar a coleta de Usages da máquina se a ação for apenas stop/kill
            if (!isStopOrKill && body.disk != 0L) {
                val diskUsageMB = srv.getUsages().disk / 1048576L // 1048576 = 1024 * 1024
                if (diskUsageMB > body.disk) {
                    srv.emitLive("error", "O uso de disco do servidor excede o limite definido.")
                    call.respond(HttpStatusCode.BadRequest, ERR_DISK_LIMIT)
                    return@post
                }
            }

            when (action) {
                "install", "start", "restart" -> {
                    if (body.image.isEmpty() || body.primaryAllocation == null || body.environment == null) {
                        call.respond(HttpStatusCode.BadRequest, ERR_MISSING_FIELDS)
                        return@post
                    }

                    // Mapeia os dados do JSON para a sua estrutura StartData
                    val startData = StartData(
                        image = body.image,
                        memory = body.memory,
                        cpu = body.cpu,
                        environment = body.environment,
                        primaryAllocation = body.primaryAllocation,
                        additionalAllocation = body.additionalAllocation,
                        disk = body.disk,
                        core = StartCore(
                            installScript = body.core.installScript,
                            installImage = body.core.installImage,
                            installEntrypoint = body.core.installEntrypoint,
                            startupCommand = body.core.startupCommand,
                            startupScript = body.core.startupScript,
                            dockerEntrypoint = body.core.dockerEntrypoint,
                            stopCommand = body.core.stopCommand,
                            configSystem = body.core.configSystem,
                            startupParser = body.core.startupParser,
                            rootAcess = body.core.rootAcess,
                            maintainable = body.core.maintainable
                        )
                    )

                    when (action) {
                        "start" -> {
                            if (!currentStatus.equals("stopped", ignoreCase = true)) {
                                srv.emitLive("error", "Servidor já em operação.")
                            } else {
                                serverActionScope.launch { srv.start(startData) }
                            }
                        }
                        "restart" -> {
                            if (!srv.isRestarting) {
                                srv.isRestarting = true
                                srv.kill()
                                serverActionScope.launch {
                                    delay(1000.milliseconds)
                                    srv.start(startData)
                                }
                            }
                        }
                        "install" -> {
                            if (!currentStatus.equals("stopped", ignoreCase = true)) {
                                call.respond(HttpStatusCode.BadRequest, ERR_NOT_STOPPED)
                                return@post
                            }
                            serverActionScope.launch { srv.install(startData, skipPull = false, keepAlive = false) }
                        }
                    }
                }

                "stop" -> {
                    if (!srv.isStopping) {
                        srv.isStopping = true
                        srv.emitLive("status", "Servidor marcado como desligando...")

                        if (body.command == "^C" || body.command == "^K") {
                            serverActionScope.launch { srv.kill() }
                        } else {
                            serverActionScope.launch { srv.sendCommand(body.command) }
                        }
                    }
                }

                "kill" -> serverActionScope.launch { srv.kill() }

                "command" -> serverActionScope.launch { srv.sendCommand(body.command) }

                else -> {
                    call.respond(HttpStatusCode.BadRequest, ERR_INVALID_ACTION)
                    return@post
                }
            }

            // Sucesso!
            call.respond(HttpStatusCode.OK, SUCCESS_RESPONSE)
        }
    }
}

File: http/src/Plume/Routes/Server/ActionsRoute.cs
using System;
using System.Collections.Generic;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Routing;
using Plume.Configuration;
using Plume.Database;
using Plume.Http;
using Plume.Server;
using Plume.Server.Extending;
using Plume.Types;

namespace Plume.Http.Routes.Server;

public static class ActionsRoute
{
    // Estrutura de DTOs para o payload (ActionPayload e ActionPayloadCore)
    public record ActionPayloadCore(
        string InstallScript = "",
        string InstallImage = "",
        string InstallEntrypoint = "",
        string StartupCommand = "",
        string StartupScript = "",
        string DockerEntrypoint = "",
        string StopCommand = "",
        JsonElement? ConfigSystem = null,
        JsonElement? StartupParser = null,
        long? RootAcess = null,
        long? Maintainable = null
    );

    public record ActionPayload(
        string Token = "",
        string ServerId = "",
        string UserUuid = "",
        string Action = "",
        string Command = "",
        int Memory = 0,
        int Cpu = 0,
        long Disk = 0,
        string Image = "",
        JsonElement? Environment = null,
        AllocationData? PrimaryAllocation = null,
        List<AllocationData>? AdditionalAllocation = null,
        ActionPayloadCore? Core = null
    );

    public static void Register(RouteGroupBuilder group)
    {
        group.MapPost("/servers/action", async (ActionPayload body, HttpContext context) =>
        {
            var result = await HandleAction(body);
            await result.ExecuteAsync(context);
        });
    }

    private static async Task<IResult> HandleAction(ActionPayload body)
    {
        // Validações Básicas
        if (string.IsNullOrEmpty(body.ServerId) || !ConfigManager.ValidateServerID(body.ServerId))
            return Reply.Json(new { error = "Invalid serverId" }, statusCode: 400);

        if (string.IsNullOrEmpty(body.UserUuid) || !ConfigManager.ValidateUUID(body.UserUuid))
            return Reply.Json(new { error = "Invalid userUuid" }, statusCode: 400);

        if (body.Token != ConfigManager.GlobalConfig.Remote.Token)
            return Reply.Json(new { error = "Invalid token" }, statusCode: 403);

        var srv = Manager.Get(body.ServerId);
        if (srv == null)
            return Reply.Json(new { error = "Server not found" }, statusCode: 404);

        var currentStatus = srv.GetStatus();
        if (currentStatus.Equals("installing", StringComparison.OrdinalIgnoreCase))
        {
            srv.EmitLive("error", "O servidor está atualmente em instalação. Tente novamente mais tarde.");
            return Reply.Json(new { error = "server is installing" }, statusCode: 400);
        }

        var serverDb = DatabaseManager.Db.GetServer(body.ServerId);
        if (serverDb == null)
            return Reply.Json(new { error = "Server not found in database" }, statusCode: 404);

        // Sincroniza limite de disco no DB e memória
        if (serverDb.Disk != body.Disk)
        {
            DatabaseManager.Db.SaveServer(serverDb with { Disk = body.Disk });
            srv.DiskLimitMb = body.Disk;
        }

        // Validação de limite de disco
        bool isStopOrKill = body.Action == "kill" || body.Action == "stop";
        if (!isStopOrKill && body.Disk != 0L)
        {
            long diskUsageMB = srv.GetUsages().Disk / 1048576L;
            if (diskUsageMB > body.Disk)
            {
                srv.EmitLive("error", "O uso de disco do servidor excede o limite definido.");
                return Reply.Json(new { error = "Disk usage exceeds the limit" }, statusCode: 400);
            }
        }

        // Processamento de Ações
        switch (body.Action.ToLower())
        {
            case "start":
            case "restart":
            case "install":
                if (body.Core == null || string.IsNullOrEmpty(body.Image) || body.PrimaryAllocation == null ||
                    body.Environment == null)
                    return Reply.Json(new { error = "Missing required fields" }, statusCode: 400);

                var startData = new StartData
                {
                    Image = body.Image,
                    Memory = body.Memory,
                    Cpu = body.Cpu,
                    Environment = body.Environment,
                    PrimaryAllocation = body.PrimaryAllocation,
                    AdditionalAllocation = body.AdditionalAllocation ?? new List<AllocationData>(),
                    Disk = body.Disk,
                    Core = new StartCore
                    {
                        InstallScript = body.Core.InstallScript,
                        InstallImage = body.Core.InstallImage,
                        InstallEntrypoint = body.Core.InstallEntrypoint,
                        StartupCommand = body.Core.StartupCommand,
                        StartupScript = body.Core.StartupScript,
                        DockerEntrypoint = body.Core.DockerEntrypoint,
                        StopCommand = body.Core.StopCommand,
                        ConfigSystem = body.Core.ConfigSystem,
                        StartupParser = body.Core.StartupParser,
                        RootAcess = body.Core.RootAcess,
                        Maintainable = body.Core.Maintainable
                    }
                };

                if (body.Action.Equals("start", StringComparison.OrdinalIgnoreCase))
                {
                    if (!currentStatus.Equals("stopped", StringComparison.OrdinalIgnoreCase))
                        srv.EmitLive("error", "Servidor já em operação.");
                    else
                        _ = Task.Run(() => srv.StartAsync(startData));
                }
                else if (body.Action.Equals("restart", StringComparison.OrdinalIgnoreCase))
                {
                    if (!srv.IsRestarting)
                    {
                        srv.IsRestarting = true;
                        _ = Task.Run(async () =>
                        {
                            await srv.KillAsync();
                            await Task.Delay(1000);
                            await srv.StartAsync(startData);
                        });
                    }
                }
                else // install
                {
                    if (!currentStatus.Equals("stopped", StringComparison.OrdinalIgnoreCase))
                        return Reply.Json(new { error = "server not stopped" }, statusCode: 400);

                    _ = Task.Run(() => srv.InstallAsync(startData, false, false));
                }

                break;

            case "stop":
                if (!srv.IsStopping)
                {
                    srv.IsStopping = true;
                    srv.EmitLive("status", "Servidor marcado como desligando...");
                    if (body.Command == "^C" || body.Command == "^K")
                        _ = Task.Run(() => srv.KillAsync());
                    else
                        _ = Task.Run(() => srv.SendCommandAsync(body.Command));
                }

                break;

            case "kill":
                _ = Task.Run(() => srv.KillAsync());
                break;

            case "command":
                _ = Task.Run(() => srv.SendCommandAsync(body.Command));
                break;

            default:
                return Reply.Json(new { error = "Invalid action" }, statusCode: 400);
        }

        return Reply.Json(new { status = "success" });
    }
}